aws_lambda_builders/workflows/dotnet_clipackage/actions.py

```diff
@@ -28,15 +31,30 @@ def __init__(self, subprocess_dotnet):
         self.subprocess_dotnet = subprocess_dotnet
 
     def execute(self):
-        try:
-            LOG.debug("Installing Amazon.Lambda.Tools Global Tool")
-            self.subprocess_dotnet.run(["tool", "install", "-g", "Amazon.Lambda.Tools", "--ignore-failed-sources"])
-        except DotnetCLIExecutionError as ex:
-            LOG.debug("Error installing probably due to already installed. Attempt to update to latest version.")
+        # run Amazon.Lambda.Tools update in sync block in case build is triggered in parallel
+        with GlobalToolInstallAction.__lock:
+            LOG.debug("Entered synchronized block for updating Amazon.Lambda.Tools")
+
+            # check if Amazon.Lambda.Tools updated recently
+            if GlobalToolInstallAction.__tools_installed:
+                LOG.info("Skipping to update Amazon.Lambda.Tools install/update, since it is updated recently")
+                return
+
             try:
-                self.subprocess_dotnet.run(["tool", "update", "-g", "Amazon.Lambda.Tools", "--ignore-failed-sources"])
+                LOG.debug("Installing Amazon.Lambda.Tools Global Tool")
+                self.subprocess_dotnet.run(["tool", "install", "-g", "Amazon.Lambda.Tools", "--ignore-failed-sources"])
+                GlobalToolInstallAction.__tools_installed = True
             except DotnetCLIExecutionError as ex:
-                raise ActionFailedError("Error configuring the Amazon.Lambda.Tools .NET Core Global Tool: " + str(ex))
+                LOG.debug("Error installing probably due to already installed. Attempt to update to latest version.")
+                try:
+                    self.subprocess_dotnet.run(
```

Comment:
Not sure whether it is a good practice to control the flow inside the exception handling or not. Is there away to check if it is already installed, then update otherwise, install?

Reply:
You are right, but I also dotnet cli doesn't have an option to say install or update if it is already installed.

Reply:
Have tried calling `update` only? It seems that is uninstall then reinstalls the tool: https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-tool-update

Reply:
I tried to run `update` only after uninstalling the tool and it worked. But I want to be cautious about what we are changing here since this is been already used widely. Maybe older version (or in any other environments) of dotnet cli might not be working as it is working right now.